### ISSUE TITLE
Port LRU improvements from 'old salsa' to salsa 2022 crate

### ISSUE THREAD

**Issue by nikomatsakis** [salsa-rs/salsa#365]
The function ingredient in salsa 2022 was based off of the code from "old salsa", but there were also some improvements made after the fork. We had to do a more detailed comparison (#366), but one improvement in particular was https://github.com/salsa-rs/salsa/pull/298, which modified the behavior when LRU eviction occurs -- instead of losing all the dependencies, we recreate the memo with the same dependencies but a different value.

**Comment by nikomatsakis** [comment]
Here are some mentoring instructions.

### Code in old salsa

Let's first trace the code in old salsa. [When an LRU overflow occurs](https://github.com/salsa-rs/salsa/blob/8dfc578edc2c62cef84c49da247dc199b50740ae/src/derived/fetch.rs#L30-L32), the `evict` method is called:

https://github.com/salsa-rs/salsa/blob/8dfc578edc2c62cef84c49da247dc199b50740ae/src/derived/fetch.rs#L106-L108

which invokes the [`evict` method on `memo_map`](https://github.com/salsa-rs/salsa/blob/8dfc578edc2c62cef84c49da247dc199b50740ae/src/derived/memo.rs#L33-L61), which (crucially) replaces the memo with a version that no value:

https://github.com/salsa-rs/salsa/blob/8dfc578edc2c62cef84c49da247dc199b50740ae/src/derived/memo.rs#L52-L58

### Eviction in salsa 2022

The salsa 2022 code triggers eviction [here](https://github.com/salsa-rs/salsa/blob/master/components/salsa-2022/src/function/fetch.rs#L26-L28), which leads to calling [`evict`](https://github.com/salsa-rs/salsa/blob/8dfc578edc2c62cef84c49da247dc199b50740ae/components/salsa-2022/src/function/fetch.rs#L96-L115), which invokes `delete_memo`:

https://github.com/salsa-rs/salsa/blob/8dfc578edc2c62cef84c49da247dc199b50740ae/components/salsa-2022/src/function/fetch.rs#L110-L112

`delete_memo` is also used elsewhere, and it has a very simple job of removing the memo from the map:

https://github.com/salsa-rs/salsa/blob/8dfc578edc2c62cef84c49da247dc199b50740ae/components/salsa-2022/src/function/delete.rs#L9-L19

### What we need to do

The old PR replaced the "delete"-style machinery with "evict". We need to keep the deletion and just add the `evict` method to the salsa 2022 memo map as an alternative. The memo map is defined here:

https://github.com/salsa-rs/salsa/blob/8dfc578edc2c62cef84c49da247dc199b50740ae/components/salsa-2022/src/function/memo.rs#L11-L14

...and you can then modify the [`evict` function](https://github.com/salsa-rs/salsa/blob/8dfc578edc2c62cef84c49da247dc199b50740ae/components/salsa-2022/src/function/fetch.rs#L96-L115) to use that newly added method, instead of calling `delete_memo`.

**Comment by OLUWAMUYIWA** [comment]
Hi Niko. Id love to work on this. Thanks!

**Comment by macovedj** [comment]
Oof @OLUWAMUYIWA not sure what to do.  I think I was working on this while you decided to pick it up.  I guess I should have commented that I was interested in it as well.  There's a PR that hasn't been reviewed yet that attempts to resolve this that I opened.  Sorry if I accidentally stepped on your toes... your comment wasn't there when I started the implementation, and I didn't see it until I finished.
